[MLIR][OpenMP] Add OpenMPToLLVMIRTranslation support for is_device_ptr

TIFitis opened the pull request:
This PR adds support for the OpenMP `is_device_ptr` clause in the MLIR to LLVM IR translation for target regions. The `is_device_ptr` clause allows device pointers (allocated via OpenMP runtime APIs) to be used directly in target regions without implicit mapping.

ergawy commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -332,10 +332,7 @@ static LogicalResult checkImplementationStatus(Operation &op) {
-    if (!op.getIsDevicePtrVars().empty())
-      result = todo("is_device_ptr");
-  };
+  auto checkIsDevicePtr = [](auto, LogicalResult &) {};
```
nit: I think we can remove this lambda altogether.

ergawy commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -3996,6 +3993,8 @@ static void collectMapDataFromMapOperands(
     auto mapType = convertClauseMapFlags(mapOp.getMapType());
     auto mapTypeAlways = llvm::omp::OpenMPOffloadMappingFlags::OMP_MAP_ALWAYS;
+    bool isDevicePtr = (mapOp.getMapType() & omp::ClauseMapFlags::storage) ==
+                       omp::ClauseMapFlags::storage;
```
nit:
```suggestion
    bool isDevicePtr = (mapOp.getMapType() & omp::ClauseMapFlags::storage);
```

ergawy reviewed (commented):
Thanks Akash. Just a few stop-by comments, I hope that's ok with you.

ergawy commented on flang/lib/Lower/OpenMP/ClauseProcessor.cpp:
```diff
@@ -1146,7 +1146,8 @@ bool ClauseProcessor::processIsDevicePtr(
         mlir::Location location = converter.genLocation(source);
         mlir::omp::ClauseMapFlags mapTypeBits =
-            mlir::omp::ClauseMapFlags::storage;
+            mlir::omp::ClauseMapFlags::return_param |
```
Why is the `return_param` flag needed here?

ergawy reviewed (commented):
Thanks Akash. Just a few comments about mapping flags. Also, can you please add an MLIR to LLVM test?

TIFitis committed:
[MLIR][OpenMP] Add OpenMPToLLVMIRTranslation support for is_device_ptr

This PR adds support for the OpenMP is_device_ptr clause in the MLIR to LLVM IR translation for target regions. The is_device_ptr clause allows device pointers (allocated via OpenMP runtime APIs) to be used directly in target regions without implicit mapping.

TIFitis commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -3875,6 +3870,10 @@ convertClauseMapFlags(omp::ClauseMapFlags mlirFlags) {
 
+  // No dedicated LLVM runtime flag for is_device_ptr; handled separately.
+  if (mapTypeToBool(omp::ClauseMapFlags::is_device_ptr))
+    mapType |= llvm::omp::OpenMPOffloadMappingFlags::OMP_MAP_RETURN_PARAM;
```
Sorry, fixed.

TIFitis commented:
> Thanks Akash. Just a few comments about mapping flags. Also, can you please add an MLIR to LLVM test?

Thanks for the review, I've added an MLIR test although it checks the same thing as the flang/test/Integration/OpenMP/map-types-and-sizes.f90 test.

TIFitis committed:
Update test.

ergawy reviewed (approved):
Thanks for updating the test Akash. LGTM. Please wait for @abhinavgaba's approval as well first.

abhinavgaba reviewed (commented):
The updated test looks good. Thanks for the changes!

On an unrelated note, for future reference, to update a PR locally, it's usually better to do `git fetch origin main && git merge origin/main && git push origin HEAD`.

Doing a `rebase` followed by a `git push --force`  clobbers the history in the PR and breaks GitHub's "see changes since last review" feature.

One scenario where a rebase + force-push can be useful is to update a branch while it still doesn't have a corresponding PR, or if the PR is in a draft stage and no reviewer has looked at it.

TIFitis commented:
> The updated test looks good. Thanks for the changes!

Thanks.

> On an unrelated note, for future reference, to update a PR locally, it's usually better to do `git fetch origin main && git merge origin/main && git push origin HEAD`.
> 
> Doing a `rebase` followed by a `git push --force` clobbers the history in the PR and breaks GitHub's "see changes since last review" feature.
> 
> One scenario where a rebase + force-push can be useful is to update a branch while it still doesn't have a corresponding PR, or if the PR is in a draft stage and no reviewer has looked at it.

Thanks for the suggestion, I'll give this a go next time.

I resorted to rebase+force push as this PR depends on #169331 and simply adding a change to this confuses GH and it starts treating them both as separate PRs and would have included changes from the base PR as part of this PR which I wanted to avoid. I don't a clean way to create PR stacks in GH like we used to be able to do in Phabricator.
